seb_openedx/middleware.py

```diff
@@ -1,27 +1,130 @@
 """ Middleware for seb_openedx """
 
 import sys
+import inspect
+from django.http import HttpResponseNotFound
 from django.utils.deprecation import MiddlewareMixin
+from django.utils.six.moves import filter as sixfilter  # pylint: disable=import-error
 from django.conf import settings
 from opaque_keys.edx.keys import CourseKey
-from seb_openedx.edxapp_wrapper.edxmako_module import render_to_response
+from web_fragments.fragment import Fragment
+from seb_openedx.edxapp_wrapper.edxmako_module import render_to_response, render_to_string
+from seb_openedx.edxapp_wrapper.get_courseware_module import get_courseware_module
 from seb_openedx.permissions import get_enabled_permission_classes
+from seb_openedx.seb_courseware_index import SebCoursewareIndex
+from seb_openedx.edxapp_wrapper.get_chapter_from_location import get_chapter_from_location
+
+SEB_WHITELIST_PATHS = getattr(settings, 'SEB_WHITELIST_PATHS', [])
+SEB_BLACKLIST_CHAPTERS = getattr(settings, 'SEB_BLACKLIST_CHAPTERS', [])
 
 
 class SecureExamBrowserMiddleware(MiddlewareMixin):
     """ Middleware for seb_openedx """
 
-    # pylint: disable=inconsistent-return-statements
     def process_view(self, request, view_func, view_args, view_kwargs):
         """ Start point of to d4etermine cms or lms """
+        courseware = get_courseware_module()
         course_key_string = view_kwargs.get('course_key_string') or view_kwargs.get('course_id')
         course_key = CourseKey.from_string(course_key_string) if course_key_string else None
+        access_denied = False
+
         if course_key:
+            # By default is all denied
+            access_denied = True
+
+            if self.is_whitelisted_view(request, course_key):
+                # First: Broad white-listing
+                access_denied = False
+
+            if self.is_blacklisted_chapter(request, course_key):
+                # Second: Granular white-listing
+                access_denied = True
+
             active_comps = get_enabled_permission_classes()
             for permission in active_comps:
                 if permission().check(request, course_key):
-                    return
-            return render_to_response('seb-403.html', status=403)
+                    access_denied = False
+
+        if access_denied:
```

Comment:
this part, where we answer once we have decided that the view needs to be denied, could be refactored into a function. Here we will be doing more things like putting a permanent mark on the user or reporting the issue somewhere

Reply:
Ok